### PR TITLE
Popover refactor + Enhancements

### PR DESCRIPTION
**Pull request by will-hitchcock**

### CODE REVIEW

**Comment by will-hitchcock** [comment]
[(3) Popover Enhancements](https://trello.com/c/cJEQM7z7/1771-3-popover-enhancements)

File: src/components/Popover/Popover.js [hunk 1]
```diff
@@ -21,15 +20,23 @@ export class Popover extends React.Component {
...
-     * Optional styles to add to the checkbox.
+    * The width of the popover, in any unit supported in css
+    */
+    width: PropTypes.string,
+    /**
+     * Optional class to add to the popover.
      */
     optClass: PropTypes.string,
+    /**
+     * Optional class to add to the popover. (DEPRECATED, use className instead)
```
Comment (samageloff) [inline src/components/Popover/Popover.js]: Swap this line w/ line 33

File: src/components/Popover/Popover.js [hunk 2]
```diff
@@ -38,84 +45,63 @@ export class Popover extends React.Component {
...
+  state = {
+    position: this.props.defaultPosition
   }
 
   componentWillReceiveProps = nextProps => {
     if (nextProps.showing) {
-      const popoverRect = this._popoverElement.getBoundingClientRect()
-      const popoverWrapper = this._popoverWrapper.getBoundingClientRect()
+      this._parentRect = this._popoverWrapper.getBoundingClientRect()
+      this._boudingRect = this._popoverElement.getBoundingClientRect()
```
Comment (fieldsco) [inline src/components/Popover/Popover.js]: misspelling: `bouding`

File: src/components/Popover/Popover.js [hunk 3]
```diff
@@ -38,84 +45,63 @@ export class Popover extends React.Component {
...
-        this.setState({ position: 'top' })
-      } else {
-        this.setState({ position: 'bottom' })
-      }
-    } else {
-      // We need to set this to the bottom each time, so that when we're subtracting the popoverRect.bottom
-      // position from the innerHeight of the window, the reference is consistent (thus not producing 'bottom' position)
-      // erroneously, which can happen from time to time
-      this.setState({ position: 'bottom' })
+  dynamicResposition = defaultPosition => {
```
Comment (fieldsco) [inline src/components/Popover/Popover.js]: misspelled method name

File: src/components/Popover/Popover.js [hunk 4]
```diff
@@ -38,84 +45,63 @@ export class Popover extends React.Component {
...
+      this._parentRect = this._popoverWrapper.getBoundingClientRect()
+      this._boudingRect = this._popoverElement.getBoundingClientRect()
 
-      // When the height of the popover + the distance of the clicked target is
-      // greater than the distance from the clicked target to the top of the browser,
-      // it's too close to the top, so set the position to 'bottom'.
-      // This enables the user to view the full popover, as they can scroll
-      // down but not up, when content is cut-off.
-      const isTooCloseToTop = popoverRect.height + popoverWrapper.height > popoverWrapper.top
+      this.setState({ position: this.dynamicResposition(nextProps.defaultPosition) })
```
Comment (samageloff) [inline src/components/Popover/Popover.js]: `dynamicReposition`

File: src/components/Popover/Popover.js [hunk 5]
```diff
@@ -38,84 +45,63 @@ export class Popover extends React.Component {
...
-      }
-    } else {
-      // We need to set this to the bottom each time, so that when we're subtracting the popoverRect.bottom
-      // position from the innerHeight of the window, the reference is consistent (thus not producing 'bottom' position)
-      // erroneously, which can happen from time to time
-      this.setState({ position: 'bottom' })
+  dynamicResposition = defaultPosition => {
+    switch (defaultPosition) {
+      case 'top':
+        if (this._boudingRect.top < 0) return 'bottom'
```
Comment (samageloff) [inline src/components/Popover/Popover.js]: `_boundingRect`

File: src/components/Popover/Popover.js [hunk 6]
```diff
@@ -38,84 +45,63 @@ export class Popover extends React.Component {
...
+        break
+      case 'bottom':
+        if (this._boudingRect.bottom > window.innerHeight) return 'top'
```
Comment (samageloff) [inline src/components/Popover/Popover.js]: `_boundingRect`

File: src/components/Popover/Popover.js [hunk 7]
```diff
@@ -38,84 +45,63 @@ export class Popover extends React.Component {
...
+        break
+      case 'left':
+        if (this._boudingRect.left < 0) return 'right'
```
Comment (samageloff) [inline src/components/Popover/Popover.js]: `_boundingRect`

File: src/components/Popover/Popover.js [hunk 8]
```diff
@@ -38,84 +45,63 @@ export class Popover extends React.Component {
...
+        break
+      case 'right':
+        if (this._boudingRect.right > window.innerWidth) return 'left'
```
Comment (samageloff) [inline src/components/Popover/Popover.js]: `_boundingRect`

**Pull request merged**